supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala

```diff
@@ -415,36 +477,83 @@ class TesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
       }
     } yield statusWithLog
 
-  private def queryStatusAsync(handle: StandardAsyncPendingExecutionHandle): Future[TesRunStatus] =
-    makeRequest[MinimalTaskView](HttpRequest(uri = s"$tesEndpoint/${handle.pendingJob.jobId}?view=MINIMAL")) map {
-      response =>
-        val state = response.state
-        state match {
-          case s if s.contains("COMPLETE") =>
-            jobLogger.info(s"Job ${handle.pendingJob.jobId} is complete")
-            Complete
+  private def queryStatusAndCostData(handle: StandardAsyncPendingExecutionHandle,
+                                     fetchCostData: Boolean
+  ): Future[TesRunStatus] =
+    if (fetchCostData) {
+      makeRequest[Task](HttpRequest(uri = s"$tesEndpoint/${handle.pendingJob.jobId}?view=FULL")) map { response =>
+        val fetchedCostData = for {
+          responseLogs <- response.logs
+          startTime <- responseLogs.head.start_time
+          vmCost <- responseLogs.head.metadata.flatMap(_.get("vm_price_per_hour_usd"))
+        } yield (startTime, vmCost)
+
+        fetchedCostData match {
+          case Some(data) =>
+            val costData = Option(TesVmCostData(data._1, data._2))
+            val state = response.state
+            val metadata = Map(
+              CallMetadataKeys.TaskStartTime -> data._1,
+              CallMetadataKeys.VmCostUsd -> data._2
+            )
+            tellMetadata(metadata)
+            getTesStatus(state, costData, handle.pendingJob.jobId)
+          case None => getTesStatus(response.state, Option.empty, handle.pendingJob.jobId)
+        }
+      }
+    } else {
+      makeRequest[MinimalTaskView](HttpRequest(uri = s"$tesEndpoint/${handle.pendingJob.jobId}?view=MINIMAL")) map {
+        response =>
+          val state = response.state
+          getTesStatus(Option(state), Option.empty, handle.pendingJob.jobId)
+      }
+    }
+
+  private def getTesStatus(state: Option[String], withCostData: Option[TesVmCostData], jobId: String): TesRunStatus = {
+    var data: Option[TesVmCostData] = Option.empty
+    if (withCostData.isDefined) {
+      data = withCostData
+      for {
+        costData <- withCostData
+        vmCost = costData.vmCost
+      } yield vmCost
```

Comment:
I'm a little confused about what's happening here - what's the purpose of `data`? Why can't we use `withCostData` directly? Also, it looks like this `for` comprehension doesn't have any effect.